### ISSUE TITLE
Swerve setup guide or walkthrough

### ISSUE THREAD

**Issue by EddieDL** [CrossTheRoadElec/Phoenix6-Examples#33]
Is there any documentation about how to get all of the initial values for the swerve drive?

We have successfully used swerve drive before, but there doesn't seem to be enough information about the basic configuration to give us confidence things are right. In addition, the reality of what is happening doesn't match what we expect.

Specifically the problem we are having now is that we can't seem to figure out the steering offset position. In the past we would align all of the modules pointing forward with the bevel gear pointing to the outside of the robot. Then we would read the absolute value from the CANCoders. That does not appear to work with whatever process is being used here.

We did determine that this implementation heavily uses firmware configurations to adjust the offset and scales the motor rotations. We think that was one problem. It doesn't seem like you can get the value of the CANCoder offset without making the current offset to be 0.

Just to be clear about our issue, once we gather the offset and attempt to run things the wheels are not oriented together.  They move properly but in different directions.

Any help would be appreciated.

**Comment by bhall-ctre** [comment]
For using the Phoenix 6 Swerve API, I would recommend using the [Tuner X Swerve Project Generator](https://v6.docs.ctr-electronics.com/en/stable/docs/tuner/tuner-swerve/index.html) to set up a working robot project.

We also have a SwerveWithPathPlanner Java example in this repo to demonstrate integration with PathPlanner. Tuner X supports exporting just the TunerConstants.java file after completing project configuration (instead of generating the entire project), so you can use that to overwrite the TunerConstants.java in the SwerveWithPathPlanner example and start off that if you plan on using PathPlanner.

**Comment by EddieDL** [comment]
Okay that makes more sense. We didn't know about the Tuner X generator, we will try that tomorrow. We were just trying to manually modify the example project to match our configuration based on the Swerve API.

It might be a good idea to add a reference to the Tuner X generator in the example project and the [Swerve API](https://v6.docs.ctr-electronics.com/en/stable/docs/api-reference/mechanisms/swerve/swerve-overview.html) documentation.

Our experience is that the current Swerve API documentation didn't provide enough information to manually configure all of the values. We are going to use the generator and then compare what it creates to what we manually tried to do. That should help us identify and understand what we were doing wrong. 

Thanks for the quick response.

**Comment by daltzctr** [comment]
The Tuner Swerve docs are actually cross-linked on the page that you are referencing.

The example project does use a `TunerConstants.json`

**Comment by EddieDL** [comment]
I see where you are talking now.
![image](https://github.com/CrossTheRoadElec/Phoenix6-Examples/assets/10708940/77f52ca1-74f5-4a37-a76c-d1f3f571886d)

I still feel like it could be more prominent or stated more forcefully. Like replacing the important call out at the top of the page with. The bulk of that page is talking about the API's so it wasn't obvious to us that the API is not the only/primary usage scenario.

> Swerve functionality is only available for FRC users and is limited to Java. We recommend using the Pheonix Tuner X generation tool if you are using this functionality.

In regard to the `TunerConstants.json` file, where is that? We only found the `TunerConstants.java` which is in a folder called genrated, but we didn't know what it was generated with. Perhaps just updating the comment at the top of the `TunerConstants.java` file to say "This file was generated using Tuner X's Swerve Project Creator. https://v6.docs.ctr-electronics.com/en/stable/docs/tuner/tuner-swerve/index.html". We just thought that CTRE used an internal tool to generate the file.

**Comment by EddieDL** [comment]
Okay. We used the tool and it didn't solve all of our problems, but it gave us more confidence in what we are doing. Below is what we found. For context we are using MK4i L2 modules with Falcons, CANCoders and a Pigeon 2 all on a Canivore.

The instructions around aligning the wheels are very confusing and not relevant for our modules.
* They ask us to put screws in places that we don't think the MK4i's have
* There is a reference "rotate the module so the bevel gear is facing towards the center of the robot". In the Pheonix Tuner X documentation it says "the bevel gear faces the vertical center of the robot" but it doesn't show or explain what that means. We are 90% sure it is asking us to place the bevel gears such that they are facing each other and the wheels are parallel with the plane that runs from the front to the back of the robot.
* If you don't have screws that can hold the module in the 0 orientation it talks about using a ruler, but there is no information about how to do that. What we did last season was use a straight edge on the flat metal of the module components that mount the wheels. That ensured things were inline plenty good enough. But it isn't clear if that is what the reference to a ruler was about.

Once we got through this we found that the PID for the steering seemed weak. The steering motors didn't drive forward and back because they would stop turning before it was at the requested direction. We just moved forward and changed the P value in our version and that seems to make things work Okay.

The remaining significant problem we have is that field centric driving doesn't seem to work properly. When we use the d-pad the robot moves forward/back and we added one for left/right. When we use field centric forward/back "generally" work (though they seem to move in a small arc). Left and right seem to rotate the robot and drive at an angle. In addition, when we use the right joystick to attempt to rotate the robot, it doesn't rotate. It just steers the wheels like it should rotate but it never drives the motors.

Any help you can provide would be appreciated.

**Comment by EddieDL** [comment]
Okay we figured out our main issue. We converted all of the main variables to be meters instead of inches. But we didn't remove the conversion when the module constants were created. We also didn't do a good job of moving the values from the generator over to our version.

We assumed it was our mistake, but it was not easy to sort out. Would have been better if we started from the creator from the beginning.

The only other suggestion we would make is to make the simulator representation of the robot include points for the modules. Not sure how hard that is, but it would have made it obvious that we made a mistake.
